src/envoy/watcher.clj

```diff
@@ -0,0 +1,69 @@
+(ns envoy.watcher
+  (:require [clojure.data :refer [diff]]
+            [clojure.core.async :refer [go-loop >!! alt! chan close!]]
+            [envoy.core :as core]
+            [org.httpkit.client :as http]))
+
+(defn- index-of [resp]
+  (-> resp
+      :headers
+      :x-consul-index))
+
+(defn- read-index
+  ([path]
+   (read-index path {}))
+  ([path ops]
+  (-> (http/get path (core/with-ops ops))
+      index-of)))
+
+(defn- start-watcher
+  ([path fun stop?]
+   (start-watcher path fun stop? {}))
+  ([path fun stop? ops]
+   (let [ch (chan)
+         close-all-channels (fn []
+                              (close! ch)
+                              (close! stop?))]
+     (go-loop [index   nil
+               current (try
+                         (core/get-all path ops)
+                         (catch RuntimeException watch-error
+                           (-> (format "Error while watching %s: %s" path watch-error)
+                               prn)
+                           (close-all-channels)))]
+              (try
+                (http/get (core/recurse path)
+                          (merge (core/with-ops (merge ops
+                                                       {:index (or index (read-index path ops))}))
+                                 (core/with-auth ops))
+                          #(>!! ch %))
+                (catch Exception watch-error
+                  (-> (format "Error while watching %s: %s" path watch-error)
+                      prn)
```

Comment:
* `watch-error` => `ex`
* no need to capitalize
* since this is prn, prefix context: `[envoy watcher]`
```suggestion
                (catch Exception watch-error
                  (-> (format "[envoy watcher]: could not read latest changes from '%s' due to: %s" path ex)
                      prn)
```